### PR TITLE
Support split online softmax reductions

### LINKED ISSUE #157975: [Inductor] Online softmax disabled due to reduction split – Unexpected performance warning
### 🐛 Bug

When running a model with `F.log_softmax` followed by `F.kl_div` (for soft targets), I receive the following warning.

This happens even if `log_softmax` and `kl_div` are inside the `forward()` method and no codegen-disrupting operations are involved.

### To Reproduce

```python
import torch
import torch.nn.functional as F

logits = torch.randn(4, 10, requires_grad=True)
targets = F.softmax(torch.randn(4, 10), dim=-1)

loss = F.kl_div(F.log_softmax(logits, dim=-1), targets, reduction="batchmean")
loss.backward()


It is only a warning but maybe you can check it fast...


### Versions

PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.10 (main, May 21 2025, 21:53:29) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.11.5-1-liquorix-amd64-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.68
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1650 Ti
Nvidia driver version: 560.35.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
CPU family:                           6
Model:                                165
Thread(s) per core:                   2
Core(s) per socket:                   6
Socket(s):                            1
Stepping:                             2
Frequency boost:                      enabled
CPU(s) scaling MHz:                   62%
CPU max MHz:                          2601.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5199.98
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust sgx bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi pku ospke sgx_lc md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            192 KiB (6 instances)
L1i cache:                            192 KiB (6 instances)
L2 cache:                             1.5 MiB (6 instances)
L3 cache:                             12 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] efficientnet_pytorch==0.7.1
[pip3] intel-cmplr-lib-ur==2024.2.1
[pip3] intel-openmp==2024.2.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] optree==0.13.1
[pip3] pytorch-lightning==2.4.0
[pip3] segmentation-models-pytorch==0.3.4
[pip3] torch==2.7.0+cu126
[pip3] torchaudio==2.7.0+cu126
[pip3] torchmetrics==1.7.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.22.0+cu126
[pip3] torchviz==0.0.3
[pip3] triton==3.3.0

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184069

Enable Inductor's online softmax lowering to keep the fused max/sum reduction when reduction splitting is selected by carrying partial max/sum tuples through the split combine stage.

Fixes #157975
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
Do we have any benchmarks that show a speedup because of this ?

**Comment by jansel** [comment]
Benchmarked this on an NVIDIA B200, comparing the PR base `87c9396b6c1` to PR head `548929776bd` with compile time excluded, a fresh `TORCHINDUCTOR_CACHE_DIR` per revision, `PYTHONPATH=.`, CUDA events, and rotating inputs.

Results:

- `torch.softmax`, bf16, `[4, 1048589]`: `0.0635 ms -> 0.0549 ms` (`1.16x`). Triton kernels: `5 -> 3`.
- `F.log_softmax`, fp32, `[4, 1048576]`: `0.0568 ms -> 0.0519 ms` (`1.09x`). Triton kernels: `5 -> 3`.
- `F.kl_div(F.log_softmax(...), ..., reduction="batchmean")` fwd+bwd, fp32, `[4, 1048576]`, reused grad buffers: `0.3226 ms -> 0.3099 ms` (`1.04x`). Triton kernels: `7 -> 5`.

So yes, the split online-softmax path speeds up the direct softmax/log_softmax cases, with a smaller end-to-end win for the KL backward case on this GPU.

**Review by shunting314 — APPROVED** [review]

**Review by eellison — APPROVED** [review]
